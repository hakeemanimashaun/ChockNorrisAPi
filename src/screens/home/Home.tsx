import { useNavigation } from "@react-navigation/native";
import React, { useEffect, useRef, useState } from "react";
import {
  BaseView,
  Button,
  MediumText,
  ParagraphText,
  ScrollView,
  Spacer,
  ViewContainer,
} from "../../components/generalStyled";
import { HomeScreenNavigationProp } from "../../navigation/Types";
import axios from "axios";
import { FlatList, Keyboard } from "react-native";
import { baseUrl } from "../../utils/BaseUrl";
import { RegularText } from "../../components/generalStyled";
import { ClickableWrapper, ViewButton } from "./components/styled";
import Search from "./components/SearchComponent";

export type data = [string];

const Home: () => JSX.Element = () => {
  const navigation = useNavigation<HomeScreenNavigationProp>();

  const [data, setData] = useState<data>();
  const [searchData, setSearchData] =
    useState<{ value: string; id: string }[]>();
  const [searchText, setSearchText] = useState<string>();
  const [showCartegories, setShowCartegories] = useState<boolean>(true);
  const [isError, setIsError] = useState<boolean>(false);
  const [isSuccess, setIsSuccess] = useState<boolean>(false);
  const searchQuery = async () => {
    await axios({
      method: "get",
      url: `${baseUrl}jokes/search?query=${searchText}`,
    })
      .then((response) => {
        setSearchData(response.data.result);
        setIsError(false);
        setIsSuccess(true);
      })
      .catch((error) => {
        console.log(error, "error");
        setIsError(true);
        setIsSuccess(false);
      });
  };

  useEffect(() => {
    axios({
      method: "get",
      url: `${baseUrl}jokes/categories`,
    })
      .then((response) => {
        setData(response.data);
      })
      .catch((error) => {
        console.log(error);
      });
  }, []);
  const clearSearch = () => {
    setShowCartegories(!showCartegories);
    setSearchText("");
  };
  const searchJokes = () => {
    setShowCartegories(!showCartegories);
    Keyboard.dismiss();
  };
  return (
    <BaseView>
      <Search
        value={searchText}
        onChangeText={(text: React.SetStateAction<string | undefined>) =>
          setSearchText(text)
        }
        onEndEditing={() => {
          searchQuery();
        }}
      />
      <Spacer height={25} />
      <Button
        onPress={() => {
          !showCartegories ? clearSearch() : searchJokes();
        }}
        testID="button"
      >
        <MediumText testID="button-text">
          {!showCartegories ? "Clear Search" : "Search Jokes"}
        </MediumText>
      </Button>
      <Spacer height={15} />
      {showCartegories ? (
        <ScrollView showsVerticalScrollIndicator={false}>
          <ViewContainer>
            {data?.map((item, index) => {
              return (
                <ViewButton
                  testID="cartegory"
                  onPress={() => {
                    navigation.navigate("Details", { item: item });
                    Keyboard.dismiss();
                  }}
                  key={index}
                >
                  <RegularText>{item}</RegularText>
                </ViewButton>
              );
            })}
          </ViewContainer>
        </ScrollView>
      ) : (
        <ViewContainer>
          {isError ? (
            <RegularText>invalid search</RegularText>
          ) : isSuccess ? (
            <FlatList
              showsVerticalScrollIndicator={false}
              data={searchData}
              renderItem={({ item }) => (
                <ClickableWrapper>
                  <ViewContainer>
                    <ParagraphText>{item.value}</ParagraphText>
                  </ViewContainer>
                </ClickableWrapper>
              )}
              keyExtractor={(item) => item.id}
            />
          ) : (
            <RegularText>search is empty</RegularText>
          )}
        </ViewContainer>
      )}
    </BaseView>
  );
};

export default Home;
