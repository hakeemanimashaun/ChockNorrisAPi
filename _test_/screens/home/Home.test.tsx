import React from "react";
import renderer, { act } from "react-test-renderer";

import Home from "../../../src/screens/home/Home";

const mockedNavigation = jest.fn();

jest.mock("@react-navigation/native", () => {
  return {
    useNavigation: () => ({
      navigate: mockedNavigation,
    }),
  };
});

describe("<Home />", () => {
  beforeEach(() => {
    mockedNavigation.mockClear();
  });

  it("renders correctly", () => {
    const tree = renderer.create(<Home />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("button press navigates", () => {
    const tree = renderer.create(<Home />);
    const button = tree.root.findByProps({ testID: "cartegory" });
    act(() => button.props.onPress());
    expect(mockedNavigation).toHaveBeenCalledWith("Details");
  });
});
