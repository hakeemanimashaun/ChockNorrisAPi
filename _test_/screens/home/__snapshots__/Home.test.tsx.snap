// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "\\"center\\"",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "boderRadius": 10,
          "borderColor": "black",
          "borderStyle": "solid",
          "borderWidth": 0.3,
          "flexDirection": "\\"row\\"",
          "height": 60,
          "justifyContent": "\\"spacebetween\\"",
          "paddingBottom": 10,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 10,
          "width": "100%",
        },
      ]
    }
  >
    <TextInput
      onChangeText={[Function]}
      onEndEditing={[Function]}
      placeholder="search keywords?"
      style={
        Array [
          Object {
            "color": "#010101",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "fontSize": 18,
            "height": 60,
            "marginRight": 10,
            "width": "100%",
          },
        ]
      }
    />
    <Text />
  </View>
  <View
    height={25}
    style={
      Array [
        Object {
          "height": 25,
        },
      ]
    }
  />
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#0066FF",
        "borderRadius": 20,
        "height": 45,
        "justifyContent": "center",
        "marginLeft": 12,
        "opacity": 1,
        "width": 150,
      }
    }
    testID="button"
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000",
            "fontSize": 18,
            "fontWeight": "500",
            "textAlign": "center",
          },
        ]
      }
      testID="button-text"
    >
      Search Jokes
    </Text>
  </View>
  <View
    height={15}
    style={
      Array [
        Object {
          "height": 15,
        },
      ]
    }
  />
  <RCTScrollView
    showsVerticalScrollIndicator={false}
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
        },
      ]
    }
  >
    <View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#fff",
              "justifyContent": "\\"center\\"",
              "paddingBottom": 10,
              "paddingLeft": 10,
              "paddingRight": 10,
              "paddingTop": 10,
            },
          ]
        }
      />
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
